content/en/blog/_posts/clientcmd-tutorial.md

```diff
@@ -0,0 +1,292 @@
+---
+layout: blog
+title: 'Uniform API server access using clientcmd'
+date: 2024-07-24
+slug: clientcmd-tutorial
+author: >
+  [Stephen Kitt](https://github.com/skitt) (Red Hat)
+---
+
+If you've ever wanted to develop a command line client for a Kubernetes API,
+especially if you've considered making your client usable as a `kubectl` plugin,
+you might have wondered how to make your client feel familiar to users of `kubectl`.
+A quick glance at the output of `kubectl options` might put a damper on that:
+"Am I really supposed to implement all those options?"
+
+Fear not, others have done a lot of the work involved for you.
+In fact, the Kubernetes provides two libraries to help you handle
+`kubectl`-style command line arguments,
+[`clientcmd`](https://pkg.go.dev/k8s.io/client-go/tools/clientcmd) and
+[`cli-runtime`](https://pkg.go.dev/k8s.io/cli-runtime)
+(which uses `clientcmd`).
+This article will show how to use the former.
+
+## General philosophy
+
+As might be expected since it's part of `client-go`,
+`clientcmd`'s ultimate purpose is to provide an instance of
+[`restclient.Config`](https://pkg.go.dev/k8s.io/client-go/rest#Config)
+that can issue requests to an API server.
+
+It follows `kubectl` semantics:
+* defaults are taken from `~/.kube` or equivalent;
+* files can be specified using the `KUBECONFIG` environment variable;
+* all of the above settings can be further overridden using command line arguments.
+
+It doesn't set up a `--kubeconfig` command line argument,
+which you might want to do to align with `kubectl`;
+you'll see how to do this
+[in the "Bind the flags" section](#bind-the-flags).
+
+## Available features
+
+`clientcmd` allows programs to handle
+
+* `kubeconfig` selection (using `KUBECONFIG`);
+* context selection;
+* namespace selection;
+* client certificates;
+* user impersonation;
+* basic authentication support (username/password).
+
+## Configuration merging
+
+In various scenarios, `clientcmd` supports _merging_ configuration settings:
+`KUBECONFIG` can specify multiple files whose contents are combined.
+This can be confusing, because settings are merged in different directions
+depending on how they are implemented.
+If a setting is defined in a map, the first definition wins,
+subsequent definitions are ignored.
+If a setting is not defined in a map, the last definition wins.
+
+When settings are retrieved using `KUBECONFIG`,
+missing files result in warnings only.
+If the user explicitly specifies a path (in `--kubeconfig` style),
+there must be a corresponding file.
+
+If `KUBECONFIG` isn't defined,
+the default configuration file, `~/.kube/config`, is used instead,
+if present.
+
+### Overall process
+
+The general usage pattern is succinctly expressed in
+[the package documentation](https://pkg.go.dev/k8s.io/client-go/tools/clientcmd):
+
+```lang-go
```

Comment:
The language identifier needs correction to enable proper syntax highlighting.
_(see [preview page (here)](https://deploy-preview-47263--kubernetes-io-main-staging.netlify.app/blog/2024/07/24/clientcmd-tutorial/), no highlighting at present)_

```suggestion
```go
```